### PR TITLE
SubscriberHistory improvements [6412]

### PR DESCRIPTION
**Pull request by MiguelCompany**
Backport of #712

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/469/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/426/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/467/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/472/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/429/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/470/)

File: src/cpp/subscriber/SubscriberHistory.cpp [hunk 1]
```diff
@@ -77,300 +90,226 @@ bool SubscriberHistory::received_change(
...
+    return receive_fn_(a_change, unknown_missing_changes_up_to);
+}
 
-    //NO KEY HISTORY
-    if (mp_subImpl->getAttributes().topic.getTopicKind() == NO_KEY)
+bool SubscriberHistory::received_change_keep_all_no_key(
+        CacheChange_t* a_change,
+        size_t unknown_missing_changes_up_to)
+{
+    // TODO(Ricardo) Check
```
Comment (LuisGP) [inline src/cpp/subscriber/SubscriberHistory.cpp]: Can this TODO comment be removed or addressed?

File: src/cpp/subscriber/SubscriberHistory.cpp [hunk 2]
```diff
@@ -77,300 +90,226 @@ bool SubscriberHistory::received_change(
...
-                if ((int32_t)m_changes.size() == m_resourceLimitsQos.max_samples)
-                    m_isHistoryFull = true;
-                logInfo(SUBSCRIBER, this->mp_subImpl->getGuid().entityId
-                    << ": Change " << a_change->sequenceNumber << " added from: "
-                    << a_change->writerGUID;);
+bool SubscriberHistory::received_change_keep_all_with_key(
+        CacheChange_t* a_change,
+        size_t /* unknown_missing_changes_up_to */ )
+{
+    // TODO(Miguel C): Should we check unknown_missing_changes_up_to as it is done in received_change_keep_all_no_key?
```
Comment (LuisGP) [inline src/cpp/subscriber/SubscriberHistory.cpp]: Can this TODO comment be removed or addressed?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/490/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/444/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/487/)

File: src/cpp/subscriber/SubscriberHistory.cpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/subscriber/SubscriberHistory.cpp]: This is a backport, and this TODO has not been addressed on the mainline.

File: src/cpp/subscriber/SubscriberHistory.cpp (hunk 2 above)
Comment (MiguelCompany) [reply src/cpp/subscriber/SubscriberHistory.cpp]: This is a backport, and this TODO has not been addressed on the mainline.

**Pull request merged**